File: Week3/Python/ArrayProblems/LongestConsecutiveSequence.py
"""
    Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.
    You must write an algorithm that runs in O(n) time.

    https://leetcode.com/problems/longest-consecutive-sequence/
"""

def longestConsecutive(nums):
    """ T(c) -> O(nlogn) + O(n), S(c) -> O(1)"""

    # empty array
    if len(nums) == 0:
        return 0

    nums.sort()

    longest_count = 1
    current_count = 1
    
    previous_num = nums[0]
    for i in range(1, len(nums)):
        current_num = nums[i]
        
        # if it is in sequence
        if current_num - 1 == previous_num:
            current_count += 1
            previous_num = current_num

        # if it is not in sequence, also not duplicate
        elif current_num != previous_num:
            current_count = 1
            previous_num = current_num
        
        longest_count = max(longest_count,  current_count)
        
    return longest_count


def longestConsecutiveEfficient(nums):
    """ T(c) -> O(n) + O(2*n), S(c) -> O(n)"""

    # to remove duplicate elements and making searching easy
    hashset = set(nums)

    current_cnt = 0
    longest_cnt = 0
        
    # iterate elements from set
    for elem in hashset:
        prev_elem = elem - 1
        
        # if current element is start of sequence
        if prev_elem not in hashset:
            current_cnt = 1
            next_elem = elem + 1
            
            # if next element is also present in set, it is in sequence
            while next_elem in hashset:
                current_cnt += 1
                next_elem = next_elem + 1
        
        # finding longest count
        longest_cnt = max(longest_cnt, current_cnt)
            
    return longest_cnt


arr = [0,3,7,2,5,8,4,6,0,1]
print(arr)
print(longestConsecutive(arr))
print(longestConsecutiveEfficient(arr))

File: Week3/Java/ArrayProblem/LongestConsecutiveSequence.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class LongestConsecutiveSequence{

/*
 * Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.
 * https://leetcode.com/problems/longest-consecutive-sequence/description/
 * T(c) -> O(nlogn) + O(n) , S(c) -> O(1)
 */

    static int longestConsecutive(int[] nums) {
        int n = nums.length;
        int currentCnt = 0;
        // keep track of longest count
        int longestCnt = 0;
        // keep track of previous number of current number sequence
        int previousNum = Integer.MIN_VALUE;
        
        // sort the array
        Arrays.sort(nums);
        
        // travarse each element and check if that is belongs to a sequence or not
        for(int i=0;i<n;i++)
        {
            int currentNum = nums[i];
            
            // it is in sequence
            if(currentNum - 1 == previousNum)
            {
                currentCnt += 1;
                previousNum = currentNum;
                
            }
            // not in sequence, not duplicate (unique number)
            else if(currentNum != previousNum){
                currentCnt = 1;
                previousNum = currentNum;
            }
            
            longestCnt = Math.max(longestCnt, currentCnt);
        }
        return longestCnt;
    }

    /* 
     * Using hash set
     * T(c) -> O(n) + O(2*n), the set will be traversed at most twice in the worst case.
     * S(c) -> O(n)
     */
    static int longestConsecutiveOptimal(int[] nums) {
        int n = nums.length;
        int currentCnt = 0;
        int longestCnt = 0;
        
        Set<Integer> set = new HashSet<Integer>();
        
        // add all elements of array to one set to avoid duplicates and searching will be easy
        for(int i=0;i<n;i++)
        {
            set.add(nums[i]);
        }
        
        // check for each element in set, if previous element present in set, that is not starting point
        for(int elem : set)
        {
            int previousElem = elem - 1;
            // if previous element not present inside set, that is the starting point
            if(!set.contains(previousElem))
            {
                currentCnt = 1;
                int nextElem = elem + 1;
                // if all consecutive next element is present in set, count it
                while(set.contains(nextElem))
                {
                    currentCnt++;
                    nextElem = nextElem + 1;
                }
                // find longest count
                longestCnt = Math.max(longestCnt, currentCnt);
            }
        }
        return longestCnt;
    }


    public static void main(String[] args) {
        int arr[] = {0,3,7,2,5,8,4,6,0,1};
        /* 0 1 2 3 4 5 6 7 8 */
        System.out.println("Longest count = " + longestConsecutive(arr));
        System.out.println("Longest count = " + longestConsecutiveOptimal(arr));
    }
}